### PR TITLE
feature: list clusters by type

### LINKED ISSUE #70: Add a filter(type=OSO/OSD) for /clusters and /clusters/auth endpoints
Currently, we are returning all clusters available when requested using GET `/clusters` and GET `clusters/auth` endpoint.

We should provide a filter of `type` to return a clusters of requested type, so that client won't have to write custom logic each time.

See https://github.com/fabric8-services/fabric8-jenkins-idler/pull/329#pullrequestreview-194464015

### PR DESCRIPTION
**Pull request by xcoulon**
optional query parameter is validated at the service
layer, and used in the GORM query if present

also, refactor some code and tests to avoid code duplication,
for Dipak's reviewing pleasure ;)

fixes #70

* Please describe what your change is about.
* What issues does it solve? (Use [autoreferencing for this](https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests))
* If you are still working on the change please prefix the pull request title with "WIP"

### CODE REVIEW

File: design/clusters.go [hunk 1]
```diff
@@ -106,21 +106,29 @@ var _ = a.Resource("clusters", func() {
 		a.Routing(
 			a.GET("/"),
 		)
+		a.Params(func() {
+			a.Param("type", d.String, "the type of the clusters to return ('OSD' or 'OSO')")
```
Comment (alexeykazakov) [inline design/clusters.go]: You don't have to implement custom validation. You can use Enum() to specify allowed values of the param. Then goa will do payload validation for your.
See an example here - https://github.com/fabric8-services/fabric8-auth/blob/3a44ba7ddd1ebe96cae360c1cecfb7f7264d69d6/design/token.go#L124

Comment (alexeykazakov) [reply design/clusters.go]: You mean JSON API Error response? It will. Check this out - https://auth.openshift.io/api/token/keys?format=unknown

Comment (xcoulon) [reply design/clusters.go]: ok, good. Let me change that, then.

File: cluster/cluster.go [hunk 2]
```diff
@@ -8,3 +13,14 @@ const (
 	// OSO the OpenShoft online type of cluster
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply cluster/cluster.go]: thanks for catching this typo, @dipak-pawar! Fixed in 6bdcdb0

File: design/clusters.go (hunk 1 above)
Comment (xcoulon) [reply design/clusters.go]: @alexeykazakov addressed in https://github.com/fabric8-services/fabric8-cluster/pull/77/commits/cbc3cc727951770e7cf36944e435e9a34d4b9bbe and https://github.com/fabric8-services/fabric8-cluster/pull/77/commits/d85720dc335064076d4925b7c8cdf7412589b6b9

**Pull request merged**